File: rules-interpreter/src/main/java/com/github/iamrenny/ruleflow/evaluators/MathAddContextEvaluator.java
package io.github.iamrenny.ruleflow.evaluators;

import io.github.iamrenny.ruleflow.RuleFlowLanguageLexer;
import io.github.iamrenny.ruleflow.RuleFlowLanguageParser.MathAddContext;
import io.github.iamrenny.ruleflow.visitors.Visitor;

import java.math.RoundingMode;

public class MathAddContextEvaluator implements ContextEvaluator<MathAddContext> {

    @Override
    public Object evaluate(MathAddContext ctx, Visitor visitor) {
        Object leftVal = visitor.visit(ctx.left);
        Object rightVal = visitor.visit(ctx.right);

        Double left = Double.valueOf(leftVal.toString());
        Double right = Double.valueOf(rightVal.toString());

        switch (ctx.op.getType()) {
            case RuleFlowLanguageLexer.ADD:
                return left + right;
            case RuleFlowLanguageLexer.MINUS:
                return left - right;
            default:
                throw new IllegalArgumentException("Operation not supported: " + ctx.op.getText());
        }
    }
}

File: rules-interpreter/src/main/kotlin/com/rappi/analang/evaluators/MathAddContextEvaluator.kt
package com.github.iamrenny.rulesflow.evaluators

import com.rappi.fraud.analang.ANALexer
import com.rappi.fraud.analang.ANAParser.MathAddContext
import com.github.iamrenny.rulesflow.visitors.Visitor
import java.math.RoundingMode

class MathAddContextEvaluator : ContextEvaluator<MathAddContext> {

    override fun evaluate(ctx: MathAddContext, visitor: Visitor): Any {
        val leftVal = visitor.visit(ctx.left)
        val rightVal = visitor.visit(ctx.right)

        val left = leftVal.toString().toBigDecimal().setScale(2, RoundingMode.DOWN)
        val right = rightVal.toString().toBigDecimal().setScale(2, RoundingMode.DOWN)
        return when (ctx.op.type) {
                ANALexer.ADD -> left.add(right)
                ANALexer.MINUS -> left.subtract(right)
                else -> throw IllegalArgumentException("Operation not supported: ${ctx.op.text}")
            }
    }

}